### PR TITLE
Issue #143: Store config version history

### LINKED ISSUE #143: Store a version history of workflows
The idea is that for any log of an instance we can see the configuration for that workflow at the point in time it was run.

Thinking the easiest way is when we run a workflow we check it's configuration, export it to yaml, and stash that in a versions table along with a hash of that configuration. We can then store the hash of the config in the data flow table, and when anything changes we unset the hash so we know it needs to be resaved. This means we only save versions of workflows that actually get run.

Also when we eventually clean up the logs of run instances we can delete any version which are no longer referenced.

### PR DESCRIPTION
**Pull request by petersistrom**
Resolves #143

### CODE REVIEW

File: classes/dataflow.php [hunk 1]
```diff
@@ -664,4 +665,22 @@ public function set_var($name, $value) {
...
+    public function save_config_version() {
+        global $DB;
+
+        $config = $this->config;
+        $configyaml = $this->config = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
+        $newconfighash = sha1($configyaml);
+
+        // Updates the stored config.
+        $DB->insert_record('tool_dataflows_versions',
+            (object) ['dataflowid' => $this->id, 'confighash' => $newconfighash, 'configyaml' => $configyaml]);
```
Comment (keevan) [inline classes/dataflow.php:682]: I know the issue mentions the configuration, but doesn't configuration also encompass name, alias, dependencies, step type, etc?

File: classes/dataflow.php [hunk 2]
```diff
@@ -664,4 +665,22 @@ public function set_var($name, $value) {
...
+     */
+    public function save_config_version() {
+        global $DB;
+
+        $config = $this->config;
+        $configyaml = $this->config = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
+        $newconfighash = sha1($configyaml);
+
+        // Updates the stored config.
+        $DB->insert_record('tool_dataflows_versions',
```
Comment (brendanheywood) [inline classes/dataflow.php:681]: Should also be an upsert? ie you could change the config back to the same as a previous state so it will have the same hash as before

File: classes/dataflow.php (hunk 1 above)
Comment (brendanheywood) [reply classes/dataflow.php:682]: I'd expect a full export and hash that then throw away the export, it should call $this->export()

File: classes/local/execution/engine.php [hunk 3]
```diff
@@ -298,6 +298,10 @@ public function finalise() {
 
             // Stores a dump of the current engine state as the finalstate of the run.
             if (isset($this->run)) {
+                // If confighash is empty save config version.
+                if ($this->dataflow->confighash === null || $this->dataflow->confighash === "") {
```
Comment (brendanheywood) [inline classes/local/execution/engine.php:302]: Might be better to move all logic around when to skip a save into save_config_version

File: classes/dataflow.php (hunk 2 above)
Comment (petersistrom) [reply classes/dataflow.php:681]: In that case should it just be a check if the record with that hash exists, if not then insert? As there is nothing to update if the hash already exists.

File: classes/dataflow.php (hunk 1 above)
Comment (petersistrom) [reply classes/dataflow.php]: Updated to $this->export()

File: classes/local/execution/engine.php (hunk 3 above)
Comment (petersistrom) [reply classes/local/execution/engine.php]: Moved logic to save_config_version

File: classes/dataflow.php [hunk 4]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
         $this->config = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
     }
+
+    /**
+     * Saves the current config to the tool_dataflows_versions table
+     *
+     */
+    public function save_config_version() {
+        if ($this->confighash === null || $this->confighash === "") {
+            global $DB;
```
Comment (keevan) [inline classes/dataflow.php:675]: Global vars should be defined at the top of the function regardless of when they are used.

File: classes/dataflow.php [hunk 5]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
         // Updates the stored config.
         $this->config = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
     }
+
+    /**
+     * Saves the current config to the tool_dataflows_versions table
+     *
```
Comment (keevan) [inline classes/dataflow.php:671]: Unnecessary whitespace here.

File: classes/dataflow.php [hunk 6]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
+     */
+    public function save_config_version() {
+        if ($this->confighash === null || $this->confighash === "") {
```
Comment (keevan) [inline classes/dataflow.php:674]:
```suggestion
        if ($this->confighash === null || $this->confighash === '') {
```

Comment (keevan) [reply classes/dataflow.php:674]: The only reason this would not work is when it's equal to zero, which is unlikely so an empty check seems more fitting here.

File: classes/dataflow.php [hunk 7]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
+            global $DB;
+
+            $config = $this->export();
+            $configyaml = Yaml::dump((array)$config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
```
Comment (keevan) [inline classes/dataflow.php:678]: Can you please ensure there is a space before and after the `(cast)` as shown in the example referenced below:

```suggestion
            $configyaml = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
```

https://www.php-fig.org/psr/psr-12/#61-unary-operators

File: classes/dataflow.php [hunk 8]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
+            $newconfighash = sha1($configyaml);
+
+            // Inserts the stored config if it is a new config.
+            if (!$DB->record_exists('tool_dataflows_versions', array('dataflowid' => $this->id, 'confighash' => $newconfighash))) {
```
Comment (keevan) [inline classes/dataflow.php:682]: Could you please use square brackets shorthand for arrays instead of the "array" language construct.

File: classes/dataflow.php [hunk 9]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
+                $DB->insert_record('tool_dataflows_versions',
+                    (object)['dataflowid' => $this->id, 'confighash' => $newconfighash, 'configyaml' => $configyaml]);
```
Comment (keevan) [inline classes/dataflow.php:684]: Can you please ensure there is a space before and after the `(cast)` as shown in the example referenced below:

https://www.php-fig.org/psr/psr-12/#61-unary-operators

File: classes/dataflow.php [hunk 10]
```diff
@@ -664,4 +665,26 @@ public function set_var($name, $value) {
...
+            }
+            // Set the config hash in dataflows table.
+            $DB->update_record('tool_dataflows', (object)['id' => $this->id, 'confighash' => $newconfighash]);
```
Comment (keevan) [inline classes/dataflow.php:687]: Can you please ensure there is a space before and after the `(cast)` as shown in the example referenced below:

https://www.php-fig.org/psr/psr-12/#61-unary-operators

File: classes/dataflow.php [hunk 11]
```diff
@@ -664,4 +665,24 @@ public function set_var($name, $value) {
         // Updates the stored config.
         $this->config = Yaml::dump((array) $config, 2, 4, Yaml::DUMP_MULTI_LINE_LITERAL_BLOCK);
     }
+    /**
```
Comment (keevan) [inline classes/dataflow.php:668]: `Expected 1 blank line after function; 0 found`

**Review by keevan — COMMENTED** [review]
You'll need to bump the version.php and add the bits to the upgrade.php script, but otherwise it seems fine.

**Review by keevan — APPROVED** [review]
